WindowsAzureMessaging/WindowsAzureMessaging/SBNotificationHub.m

```diff
@@ -378,19 +378,58 @@ - (void)deleteRegistrationWithName:(NSString *)templateName completion:(void (^)
 - (void)unregisterAllWithDeviceToken:(NSData *)deviceTokenData completion:(void (^)(NSError *))completion {
     if (deviceTokenData == nil) {
         if (completion) {
-            completion([SBNotificationHubHelper errorForNullDeviceToken]);
+            dispatch_async(dispatch_get_main_queue(), ^{
+                completion([SBNotificationHubHelper errorForNullDeviceToken]);
+            });
         }
 
         return;
     }
 
     NSString *deviceToken = [self convertDeviceToken:deviceTokenData];
-
-    SBThreadParameter *parameter = [[SBThreadParameter alloc] init];
-    parameter.deviceToken = deviceToken;
-    parameter.completion = completion;
-    parameter.isMainThread = [[NSThread currentThread] isMainThread];
-    [self performSelectorInBackground:@selector(deleteAllRegistrationThread:) withObject:parameter];
+    
+    [self retrieveAllRegistrationsWithDeviceToken:deviceToken completion:^(NSArray *registrations, NSError *error) {
+        
+        if (error) {
+            dispatch_async(dispatch_get_main_queue(), ^{
+                if (completion) {
+                    completion(error);
+                }
+            });
+            return;
+        }
+        
+        if (registrations.count > 0) {
+            dispatch_group_t group = dispatch_group_create();
+            __block NSError *innerError = nil;
+            
+            for (SBRegistration *reg in registrations) {
+                NSString *name = [SBNotificationHubHelper nameOfRegistration:reg];
+                dispatch_group_enter(group);
+                [self deleteRegistrationWithName:name completion:^(NSError *err) {
+                    innerError = err;
+                    dispatch_group_leave(group);
+                }];
+            }
+            
+            dispatch_group_notify(group, dispatch_get_main_queue(),^{
+                if (!innerError) {
+                    [self->storageManager deleteAllRegistrations];
+                }
+                
+                if (completion) {
+                    completion(innerError);
+                }
+            });
+        } else {
+            dispatch_async(dispatch_get_main_queue(), ^{
+                [self->storageManager deleteAllRegistrations];
+                if (completion) {
+                    completion(error);
+                }
```

Comment:
We have:
```objc
if (error) {
    completion ()
}
```
At the very top of the method, unless I'm mistaken, this call will be done with nil.